algorithms/linfa-clustering/src/optics/algorithm.rs

```diff
@@ -0,0 +1,498 @@
+pub struct OpticsAnalysis<'a, F: Float> {
+    /// A list of the samples in the dataset sorted and with their reachability and core distances
+    /// computed.
+    pub orderings: Vec<Sample<'a, F>>,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct Sample<'a, F: Float> {
+    /// Index of the sample in the dataset.
+    pub index: usize,
+    /// A reference to the datum.
+    pub observation: ArrayView<'a, F, Ix1>,
+    /// The reachability distance of a sample is the distance between the point and it's cluster
+    /// core or another point whichever is larger.
+    pub reachability_distance: Option<f64>,
+    /// The distance to the nth closest point where n is the minimum points to form a cluster.
+    pub core_distance: Option<f64>,
+}
+
+#[derive(Clone)]
+struct Neighbor<'a, F: Float> {
+    /// Index of the observation in the dataset
+    index: usize,
+    /// The observation
+    observation: ArrayView<'a, F, Ix1>,
+    /// The core distance, named so to avoid name clash with `Sample::core_distance`
+    c_distance: Option<FloatOrd<f64>>,
+    /// The reachability distance
+    r_distance: Option<FloatOrd<f64>>,
+}
+
+impl<'a, F: Float> Neighbor<'a, F> {
+    /// Create a new neighbor
+    fn new(index: usize, observation: ArrayView<'a, F, Ix1>) -> Self {
+        Self {
+            index,
+            observation,
+            c_distance: None,
+            r_distance: None,
+        }
+    }
+
+    /// Set the core distance given the minimum points in a cluster and the points neighbors
+    fn set_core_distance(&mut self, min_pts: usize, neighbors: &[Neighbor<'a, F>]) {
+        self.c_distance = neighbors
+            .get(min_pts - 1)
+            .map(|x| FloatOrd(self.observation.l2_dist(&x.observation).unwrap()));
+    }
+
+    /// Convert the neighbor to a sample for the user
+    fn sample(&self) -> Sample<'a, F> {
+        Sample {
+            index: self.index,
+            observation: self.observation.clone(),
+            reachability_distance: self.r_distance.map(|x| x.0),
+            core_distance: self.c_distance.map(|x| x.0),
+        }
+    }
+}
+
+impl<'a, F: Float> Eq for Neighbor<'a, F> {}
+
+impl<'a, F: Float> PartialEq for Neighbor<'a, F> {
+    fn eq(&self, other: &Self) -> bool {
+        self.r_distance == other.r_distance
+    }
+
+    fn ne(&self, other: &Self) -> bool {
+        !self.eq(other)
+    }
+}
+
+impl<'a, F: Float> PartialOrd for Neighbor<'a, F> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        self.r_distance.partial_cmp(&other.r_distance)
+    }
+}
+
+impl<'a, F: Float> Ord for Neighbor<'a, F> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.r_distance.cmp(&other.r_distance)
+    }
+}
+
+impl Optics {
+    pub fn params(min_points: usize) -> OpticsHyperParams {
+        OpticsHyperParams::new(min_points)
+    }
+}
+
+impl<'a, F: Float, D: Data<Elem = F>>
+    Transformer<&'a ArrayBase<D, Ix2>, Result<OpticsAnalysis<'a, F>>> for OpticsHyperParams
```

Comment:
Did you manage to fix the lifetime issue here? If not then I suggest you remove the `ArrayViews` from `OpticsAnalysis` and just use the indices, since that's what DBSCAN does. Also the trait here should be `Predict`, which is what's used by DBSCAN right now.

Reply:
OPTICS is more "meta" than DBSCAN, for example DBSCAN predicts clusters, whereas OPTICS generates some metadata for each data point that could be used to implement something like DBSCAN, it doesn't actually do any clustering

Reply:
Now that I think about it you're right. OPTICs returns distance info on every point so it makes more sense to use `Transformer`.